### ISSUE TITLE
New feature: could you add option to set default headers?

### ISSUE THREAD

**Issue by Martinsos** [janpantel/angular-sails#47]
angularSails supports setting of default headers, could angular-sails also support that?
Check http://balderdashy.github.io/angularSails/#/api/ngsails.$sailsSocket, under Setting Headers.
Thanks!

**Comment by TheSharpieOne** [comment]
We are trying to update everything to work with the new sails.io.js abilities while still working for older sails.io.js.  Headers is one of the things that is involved with this.  Its in the works along with full test suites for more code confidence as its a lot of changes.

**Comment by psi-4ward** [comment]
See; https://github.com/janpantel/angular-sails/blob/master/src/service/angular-sails.js#L67
You could add a request interceptor which adds your headers.

**Comment by TheSharpieOne** [comment]
That will not work because angular-sails uses sails.io.js.  sails.io.js will just ignore the headers when it makes the actual call. See: https://github.com/balderdashy/sails.io.js/blob/master/sails.io.js#L725

**Comment by dirkjan111** [comment]
Is token based Authentication currently possible with angular-sails.., it needs the authorization header?

**Comment by TheSharpieOne** [comment]
headers (and thus setting the auth header) is on the roadmap for [2.0](https://github.com/janpantel/angular-sails/tree/2.0) which is not 100% ready, but close.

**Comment by psi-4ward** [comment]
Oh rlly? No header support?

Looked into my sails project: Ive a `/login` and `/login/authSocket` route with support to authenticate the Socket. So i dont have to send credentials every request.

**Comment by ocorsvall** [comment]
Any update on the progress on this?
As I also would like to be able to send a token in the header.

**Comment by snowcxt** [comment]
+1

**Comment by tommykennedy** [comment]
+1

**Comment by TheSharpieOne** [comment]
I was going to mention [beta release of 2.0](https://github.com/janpantel/angular-sails/releases/tag/2.0.0-beta.1) and say that it has default headers... but it has been a while since I made it and looking at it now I see.... [`// TODO: default headers`](https://github.com/janpantel/angular-sails/blob/2.0/src/service/%24sails.js#L66) :(

But while I'm at it... RFC for what the default default headers should be (if any).  Angular's `$http` has [some defaults](https://docs.angularjs.org/api/ng/service/$http#setting-http-headers) mostly around `Content-Type`... not sure if that is relevant with `$sails` or not.

**Comment by TheSharpieOne** [comment]
So... there is now [2.0 beta 2](https://github.com/janpantel/angular-sails/releases/tag/2.0.0-beta.2) release.  It has default headers abilities similar to how angular's `$http` works, just use `$sailsProvider` instead of `$httpProvider` (i.e. `$sailsProvider.defaults.headers.get = {'my-header': 'some value'}`).

Let me know if you have any issues with 2.0 beta.

**Comment by AdamJessop** [comment]
I have just tested this with the following from an AuthService:

$sails.defaults.headers.common = {'Authorization' : 'Bearer ' + token};
$sails.defaults.headers.post = {'Authorization' : 'Bearer ' + token};

Checking the socket frames in chrome devtools I can see that the headers appear to be empty.

**Comment by snowcxt** [comment]
Is it possible to automatically append the http headers to the socket, which means we just have to replace the $http with $sails?

**Comment by TheSharpieOne** [comment]
@snowcxt not sure what you mean.  Something like "use the headers set in `$httpProvider.default.headers` when making `$sails` calls?  So it would be something like

`finalHeadersSent` <- `$http.default.headers.common` <- `$http.default.headers[method]` <- `$sails.default.headers.common` <- `$sails.default.headers[method]` <- `request.config.headers`

Each one extending the other to add or override the values.

That could be useful when you want to use both `$http` and `$sails` together so you don't have to duplicate setting the headers on `$httpProvider` and `$sailsProvider`

**Comment by snowcxt** [comment]
Right. And it's nice to have the "headers" parameter in the $sails function as what $http does.

**Comment by TheSharpieOne** [comment]
It is.  You can pass headers in the config and they will be merged into the defaults. Current it is set up to do something like:
`finalHeadersSent` <- `$sails.default.headers.common` <- `$sails.default.headers[method]` <- `request.config.headers`

as you can see [here](https://github.com/janpantel/angular-sails/blob/2.0.0-beta.2/src/service/%24sails.js#L80) and really [here](https://github.com/janpantel/angular-sails/blob/2.0.0-beta.2/src/util.js#L36-L58)

**Comment by TheSharpieOne** [comment]
@AdamJessop It turns out that if I don't actually update the dist files, then no one gets them :(  lol.. yeah, I forgot to run gulp to create the dist files that you use.  So 2.0.0-beta-2 dist files are actually 1.x... No fear, [2.0.0-beta.3](https://github.com/janpantel/angular-sails/releases/tag/2.0.0-beta.3) has the legit dist files now and you should be able to see the headers working! Give it a try.

@snowcxt as far as your request.  I would probably put that ability behind a flag, and if I were to do that and have users/devs set the flag, I might as well just recommend doing `$sailsProvider.default.headers = $httpProvider.default.headers` as it would be the same affect.

**Comment by jojo897** [comment]
Has this feature been implemented yet ? Doesn't look like it has, but I may be wrong ...

**Comment by ciriac** [comment]
Any update on this?
